### ISSUE TITLE
linker errors: errors undefined reference to `aws_default_allocator'

### ISSUE THREAD

**Issue by curiousjgeorge** [aws/aws-sdk-cpp#1191]
I ran into an issue using the AWS C++ SDK as a static library and building with clang. I built the project's three dependencies separately, as static libraries. I received linker errors when attempting to use the C++ library. The error messages had statements like "undefined reference to 'aws_default_allocator'" and other items that belong to aws-c-common..

To get around the issue, I had to edit aws-c-common-0.3.13's cmake/AwsCFlags.cmake file, and remove the `list(APPEND AWS_C_FLAGS -fvisibility=hidden)`. After re-compiling aws-c-common, aws-c-event-stream, and aws-cpp-sdk, my errors went away.

Does this suggest an issue with aws-c-common - should it be setting the visibility to hidden?

### What platform/OS are you using?
CentOS 7

### Which version of the SDK?
1.7.124

### What compiler are you using? what version?
clang

### Can you provide a TRACE level log? (sanitize any sensitive information)
I cannot provide a log at this time

**Comment by KaibaLopez** [comment]
hi @curiousjgeorge, I'm having trouble replicating this issue, for me the linking is working just fine following our readme. 
Can you provide the cmake arguments you used to build the deps and the sdk?

**Comment by curiousjgeorge** [comment]
If I build aws-c-common with `list(APPEND AWS_C_FLAGS -fvisibility=hidden)` in place, these are the cmake arguments I use:
```
cmake aws-c-common-0.3.13
-DCMAKE_BUILD_TYPE=Release
-DCMAKE_INSTALL_PREFIX=(redacted)/
-DCMAKE_INSTALL_LIBDIR=(redacted)/lib
-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON
```

When I build aws-sdk-cpp, to get static libraries, these are the cmake arguments I use:
```
 cmake aws-sdk-cpp-1.7.124
-DCMAKE_BUILD_TYPE=Release
-DCMAKE_INSTALL_PREFIX=(redacted)
-DCMAKE_INSTALL_LIBDIR=(redacted)/lib
-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON
'-DBUILD_ONLY=s3'
-DBUILD_DEPS=OFF
-DBUILD_SHARED_LIBS=OFF
-DCUSTOM_MEMORY_MANAGEMENT=0
```

They build fine. The problem is at the linker stage with clang when I use the libraries with a small, simple test program. The linker uses the following aws libraries:
* libaws-cpp-sdk-s3.so
* libaws-cpp-sdk-core.so
* libaws-c-event-stream.so
* libaws-checksums.so
* libaws-c-common.so


(Note: There is some other magic behind the scenes, where static libraries are packaged up into dynamic libraries, so that's why you see ".so" files instead of ".a" files.)

When linking, I get the errors like `libaws-cpp-sdk-core.so: error: undefined reference to 'aws_default_allocator'`

If I run `nm` on the libraries and grep for that symbol, this is the output (I did not include the libraries where there is not any output):
```
$ nm -A libaws-cpp-sdk-core.so | grep aws_default_alloc
                 U aws_default_allocator
$ nm -A libaws-c-common.so  | grep aws_default_alloc
000000000000d6a0 t aws_default_allocator
```

I can get it all working by removing the `list(APPEND AWS_C_FLAGS -fvisibility=hidden)` line in aws-c-common. When I ran it by the experts I know, they suggested it was a possible bug in aws-c-common.  (I am not an expert, I'm only relaying back information, hopefully I'm being accurate)
